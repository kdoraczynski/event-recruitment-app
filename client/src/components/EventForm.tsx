import React, { useCallback } from 'react';
import { Formik } from 'formik';
import { Form, InputGroup, Row, Col, Card } from 'react-bootstrap';
import eventSchema from './event.schema';
import { EventModel } from '../shared/api/eventModel';
import ProgressButton from './ProgressButton';

type EventFormProps = {
	onSubmit: (eventData: EventModel) => Promise<EventModel>
}

export const getEventFormInitialValues = (): EventModel => ({
	firstName: '',
	lastName: '',
	email: '',
	date: '',
});

const EventForm = ({ onSubmit }: EventFormProps) => {
	const handleSubmit = useCallback(async (formData, { setSubmitting, resetForm }) => {
		setSubmitting(true);
		await onSubmit(formData);
		resetForm();
		setSubmitting(false);
	}, [ onSubmit ]);

	return (
		<Card
			className="mt-3"
			body
		>
			<Formik
				initialValues={getEventFormInitialValues()}
				validationSchema={eventSchema}
				onSubmit={handleSubmit}
			>
				{
					({
						 handleSubmit,
						 isSubmitting,
						 handleChange,
						 handleBlur,
						 touched,
						 errors,
						 values
					 }) => (
						<Form
							noValidate
							onSubmit={handleSubmit}
						>
							<Form.Group
								as={Row}
								controlId="firstName"
								key="firstName"
								className=''
							>
								<Form.Label
									column
									sm="3"
								>
									First name
								</Form.Label>
								<Col sm="9">
									<Form.Control
										name="firstName"
										value={values.firstName}
										onChange={handleChange}
										onBlur={handleBlur}
										isInvalid={touched.firstName && !!errors.firstName}
									/>
									<Form.Control.Feedback
										type='invalid'
										className='text-left'
									>
										{errors.firstName}
									</Form.Control.Feedback>
								</Col>
							</Form.Group>
							<Form.Group
								as={Row}
								controlId="lastName"
								key="lastName"
							>
								<Form.Label
									column
									sm="3"
								>
									Last name
								</Form.Label>
								<Col sm="9">
									<Form.Control
										name="lastName"
										value={values.lastName}
										onChange={handleChange}
										onBlur={handleBlur}
										isInvalid={touched.lastName && !!errors.lastName}
									/>
									<Form.Control.Feedback
										type='invalid'
										className='text-left'
									>
										{errors.lastName}
									</Form.Control.Feedback>
								</Col>
							</Form.Group>
							<Form.Group
								as={Row}
								controlId="email"
								key="email"
							>
								<Form.Label
									column
									sm="3"
								>
									Email address
								</Form.Label>
								<Col sm="9">
									<InputGroup>
										<InputGroup.Prepend>
											<InputGroup.Text>@</InputGroup.Text>
										</InputGroup.Prepend>
										<Form.Control
											name="email"
											value={values.email}
											onChange={handleChange}
											onBlur={handleBlur}
											isInvalid={touched.email && !!errors.email}
										/>
										<Form.Control.Feedback
											type='invalid'
											className='text-left'
										>
											{errors.email}
										</Form.Control.Feedback>
									</InputGroup>
								</Col>
							</Form.Group>
							<Form.Group
								as={Row}
								controlId="date"
								key="date"
							>
								<Form.Label
									column
									sm="3"
								>
									Event date
								</Form.Label>
								<Col sm="9">
									<Form.Control
										type='date'
										name="date"
										value={values.date}
										onChange={handleChange}
										onBlur={handleBlur}
										isInvalid={touched.date && !!errors.date}
									/>
									<Form.Control.Feedback
										type='invalid'
										className='text-left'
									>
										{errors.date}
									</Form.Control.Feedback>
								</Col>
							</Form.Group>
							<ProgressButton
								type='submit'
								isLoading={isSubmitting}
							>
								Sign up
							</ProgressButton>
						</Form>
					)
				}
			</Formik>
		</Card>
	);
};

export default EventForm;
