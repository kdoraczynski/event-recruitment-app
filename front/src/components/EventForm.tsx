import React from 'react';
import Card from 'react-bootstrap/cjs/Card';
import { Form, InputGroup } from 'react-bootstrap';
import Row from 'react-bootstrap/Row';
import { Formik } from 'formik';
import eventSchema from './event.schema';
import { getDefault } from '../shared/api/events.api';
import Col from 'react-bootstrap/Col';
import Button from 'react-bootstrap/Button';

type Props = {
	onSubmit: (...args: any[]) => void
}

const initialValues = getDefault()

const EventForm = ({ onSubmit }: Props) => {
	return (
		<Card
			className="mt-3"
			body
		>
			<Formik
				initialValues={initialValues}
				validationSchema={eventSchema}
				onSubmit={(values, { setSubmitting }) => {
					setSubmitting(true)
					onSubmit(values)
					setSubmitting(false)
				}}
			>
				{
					({
						handleSubmit,
						isSubmitting,
						handleChange,
						handleBlur,
						touched,
						errors,
						values
					}) => (
						<Form
							noValidate
							onSubmit={handleSubmit}
						>
							<Form.Group
								as={Row}
								controlId="firstName"
								key="firstName"
							>
								<Form.Label
									column
									sm="3"
								>
									First name
								</Form.Label>
								<Col sm="9">
									<Form.Control
										name="firstName"
										value={values.firstName}
										onChange={handleChange}
										onBlur={handleBlur}
										isInvalid={!!errors.firstName}
									/>
									<Form.Control.Feedback type={'invalid'}>
										{ errors.firstName }
									</Form.Control.Feedback>
								</Col>
							</Form.Group>
							<Form.Group
								as={Row}
								controlId="lastName"
								key="lastName"
							>
								<Form.Label
									column
									sm="3"
								>
									Last name
								</Form.Label>
								<Col sm="9">
									<Form.Control
										name="lastName"
										value={values.lastName}
										onChange={handleChange}
										onBlur={handleBlur}
										isInvalid={!!errors.lastName}
									/>
									<Form.Control.Feedback type={'invalid'}>
										{ errors.lastName }
									</Form.Control.Feedback>
								</Col>
							</Form.Group>
							<Form.Group
								as={Row}
								controlId="email"
								key="email"
							>
								<Form.Label
									column
									sm="3"
								>
									Email address
								</Form.Label>
								<Col sm="9">
									<InputGroup>
										<InputGroup.Prepend>
											<InputGroup.Text>@</InputGroup.Text>
										</InputGroup.Prepend>
									<Form.Control
										name="email"
										value={values.email}
										onChange={handleChange}
										onBlur={handleBlur}
										isInvalid={!!errors.email}
									/>
									<Form.Control.Feedback type={'invalid'}>
										{ errors.email }
									</Form.Control.Feedback>
									</InputGroup>
								</Col>
							</Form.Group>
							<Form.Group
								as={Row}
								controlId="date"
								key="date"
							>
								<Form.Label
									column
									sm="3"
								>
									Event date
								</Form.Label>
								<Col sm="9">
									<Form.Control
										type='date'
										name="date"
										value={values.date}
										onChange={handleChange}
										onBlur={handleBlur}
										isInvalid={!!errors.date}
									/>
									<Form.Control.Feedback type={'invalid'}>
										{ errors.date }
									</Form.Control.Feedback>
								</Col>
							</Form.Group>
							<Button
								type="submit"
							>
								Sign up
							</Button>
						</Form>
					)
				}
			</Formik>
		</Card>
	);
};

export default EventForm;
