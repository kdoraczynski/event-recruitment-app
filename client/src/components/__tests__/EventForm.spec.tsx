import * as React from 'react';
import renderer from 'react-test-renderer';
import EventForm, { getEventFormInitialValues } from '../EventForm';

describe('EventForm', () => {
	it('getEventFormInitialValues() should return new event object', () => {
		const initialEventFormValues = getEventFormInitialValues();

		expect(initialEventFormValues).toMatchObject({
			firstName: '',
			lastName: '',
			email: '',
			date: ''
		});
	});

	it('Renders correctly', () => {
		const eventForm = renderer.create(<EventForm onSubmit={jest.fn()}/>)
			.toJSON();

		expect(eventForm).toMatchSnapshot();
	});
});
