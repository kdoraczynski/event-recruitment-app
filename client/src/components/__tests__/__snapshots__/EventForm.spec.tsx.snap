// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventForm Renders correctly 1`] = `
<div
  className="mt-3 card"
>
  <div
    className="card-body"
  >
    <form
      className=""
      noValidate={true}
      onSubmit={[Function]}
    >
      <div
        className="form-group row"
      >
        <label
          className="form-label col-form-label col-sm-3"
          htmlFor="firstName"
        >
          First name
        </label>
        <div
          className="col-sm-9"
        >
          <input
            className="form-control"
            id="firstName"
            name="firstName"
            onBlur={[Function]}
            onChange={[Function]}
            value=""
          />
          <div
            className="text-left invalid-feedback"
          />
        </div>
      </div>
      <div
        className="form-group row"
      >
        <label
          className="form-label col-form-label col-sm-3"
          htmlFor="lastName"
        >
          Last name
        </label>
        <div
          className="col-sm-9"
        >
          <input
            className="form-control"
            id="lastName"
            name="lastName"
            onBlur={[Function]}
            onChange={[Function]}
            value=""
          />
          <div
            className="text-left invalid-feedback"
          />
        </div>
      </div>
      <div
        className="form-group row"
      >
        <label
          className="form-label col-form-label col-sm-3"
          htmlFor="email"
        >
          Email address
        </label>
        <div
          className="col-sm-9"
        >
          <div
            className="input-group"
          >
            <div
              className="input-group-prepend"
            >
              <span
                className="input-group-text"
              >
                @
              </span>
            </div>
            <input
              className="form-control"
              id="email"
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              value=""
            />
            <div
              className="text-left invalid-feedback"
            />
          </div>
        </div>
      </div>
      <div
        className="form-group row"
      >
        <label
          className="form-label col-form-label col-sm-3"
          htmlFor="date"
        >
          Event date
        </label>
        <div
          className="col-sm-9"
        >
          <input
            className="form-control"
            id="date"
            name="date"
            onBlur={[Function]}
            onChange={[Function]}
            type="date"
            value=""
          />
          <div
            className="text-left invalid-feedback"
          />
        </div>
      </div>
      <button
        className="btn btn-primary"
        disabled={false}
        type="submit"
      >
        Sign up
      </button>
    </form>
  </div>
</div>
`;
